### DISCUSSION TITLE
Code coverage reporting

### DISCUSSION THREAD

**Discussion by sevenam** [thomhurst/TUnit#697] [General]
For xunit we've normally been creating code coverage reports with:

`dotnet test --collect:"XPlat Code Coverage"`

Is there a way to do this with TUnit or is this a feature that's not been added yet?

**Comment by thomhurst** [comment]
Heya @sevenam - Details are here: https://thomhurst.github.io/TUnit/docs/extensions/

**Reply by sevenam** [reply]
Thanks! I missed this when going through the documentation, sorry! It would be great to have an example of how to use it as well.
Was hoping to make it work and to maybe do a PR with an example, but I didn't quite manage to get it to work.

Added the nuget in the .csproj file:

```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <IsPackable>false</IsPackable>
    <IsTestProject>true</IsTestProject>
    <CollectCoverage>true</CollectCoverage>
    <CoverletOutputFormat>cobertura</CoverletOutputFormat>
    <CoverletOutput>$(OutputPath)coverage.cobertura.xml</CoverletOutput>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="FluentAssertions" Version="6.12.0" />
    <PackageReference Include="Microsoft.Testing.Extensions.CodeCoverage" Version="17.12.5" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="8.0.8" />
    <PackageReference Include="Microsoft.AspNetCore.TestHost" Version="8.0.8" />
    <!-- had to remove this reference as TUnit tests would not run when it was included -->
    <!--<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.11.1" />-->
    <PackageReference Include="TUnit" Version="0.1.897" />
  </ItemGroup>
```

Then tried running the tests with:
`dotnet test ProjectFolder --collect:"Code Coverage"`
and
`dotnet test ProjectFolder --collect:"XPlat Code Coverage"`

Also tried adding a runsettings file:
```
<?xml version="1.0" encoding="utf-8"?>
<RunSettings>
  <DataCollectionRunSettings>
    <DataCollectors>
      <DataCollector friendlyName="Code Coverage">
        <Configuration>
          <Format>cobertura</Format>
        </Configuration>
      </DataCollector>
    </DataCollectors>
  </DataCollectionRunSettings>
</RunSettings>
```
and pointing to this with dotnet test:
`dotnet test ProjectFolder --collect:"Code Coverage" --settings codecoverage.runsettings`

Maybe there's something obvious I missed...

**Reply by thomhurst** [reply]
The flag is `--coverage`. See here https://thomhurst.github.io/TUnit/docs/tutorial-extras/command-line-flags

**Reply by sevenam** [reply]
Great - thanks a lot! So there was something obvious I'd missed...

Got it to work with this input to **dotnet test** (I had initially missed the additional "--" that was required).

`dotnet test ProjectFolder -- --coverage --coverage-output coverage.cobertura.xml --coverage-output-format cobertura --results-directory .`

It didn't output any .xml until I included the `--results-directory` in addition to the `--coverage` flags.

**Reply by thomhurst** [reply]
I believe by default it goes into bin/Test results/coverage.xml

**Comment by gioce90** [comment]
The `--collect:"XPlat Code Coverage"` is the command for vstest driver to activate coverlet.collector, and `--collect:"Code Coverage"` instead uses the dotnet-coverage tool. 
The `--coverage` in TUnit is using this last one as default, and it's great, but there is a way to change the collector?

**Reply by thomhurst** [reply]
Coverlet isn't compatible currently so I think you're stuck with that donet collector for now.

**Comment by gioce90** [comment]
What about inline params? There is a way to put simple settings directly in the command?

For example, for years in my projects on Azure DevOps the coverage was including also the Test projects and the third-party libraries coverage... and this was a bit frustrating because our pipelines where showing a false percentage of code coverage. 
[Luckily then](https://github.com/microsoft/codecoverage/issues/20#issuecomment-1830207312) was introduced `IncludeTestAssembly=false` to be used, for example, in this way:

```
- task: DotNetCoreCLI@2
  displayName: 'dotnet test'
  inputs:
    command: 'test'
    projects: '**/*.csproj'  #or *.sln
    arguments: '--no-restore --configuration $(buildConfiguration) --collect "Code Coverage;IncludeTestAssembly=false"' 
```

This would be a convenient way to quickly pass parameters without using a XML settings file (runsettings.xml can be really frustrating sometimes).

**Reply by thomhurst** [reply]
The coverage extension isn't actually part of TUnit, it's a Microsoft tool. So if you want features for it you're better off raising it with them as I can't do anything. The docs on their extension is here https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-extensions-code-coverage

**Reply by gioce90** [reply]
Sorry, maybe it seems I was focused too much on coverage features, but the goal here is _the way how TUnit pass parameters_ to the underline MS coverage tool.

I read your [docs](https://thomhurst.github.io/TUnit/docs/tutorial-extras/command-line-flags/) and I saw:

```
--coverage-settings
XML code coverage settings
```

And maybe I'm wrong, but my worries is about that there is no way to pass quick input to the collector (like `--collect "Code Coverage;IncludeTestAssembly=false"` but could be everything else) and the only way is to use the XML settings file.

**Reply by thomhurst** [reply]
TUnit doesn't pass anything explicitly. The extension point is all owned by Microsoft and if you want to be able to pass stuff, you'd still need to raise an issue with them.

**Comment by gioce90** [comment]
What about default exclusions from coverage?

As mentioned, test assemblies and third-party libraries are usually included in coverage, which can be confusing when you need to focus on the % coverage of your codebase.

With the dotnet collector, you can exclude test assemblies with `--collect "Code Coverage;IncludeTestAssembly=false"`.
Now is also possible to exclude third-party libraries with the  `ExcludeAssembliesWithoutSources=MissingAll` [workaround](https://github.com/microsoft/codecoverage/issues/12#issuecomment-1907966480) .
In the past I asked [the introduction of a parameter](https://github.com/microsoft/codecoverage/issues/72), but with no luck.

However, parameters aside, I think excluding test assemblies and third-party libraries should be the default. _People usually want to focus on their codebase_.

So, in my opinion:
- the `--collect` command in TUnit should allow parameters, like `--collect "IncludeTestAssembly=true"` or something like that.
- test assemblies should be excluded from coverage by default. But it should be possible to include them with `IncludeTestAssembly=true`.
- third-party libraries should also be excluded from coverage by default. [Coverlet is already doing this](https://github.com/coverlet-coverage/coverlet/issues/1164). But should be possible to include them with "IncludeThirdPartyLibraries=true", or something like that.

**Reply by thomhurst** [reply]
As mentioned, the collect flags etc aren't part of TUnit so I can't physically change anything about them. You need to raise with MS 😄

**Reply by gioce90** [reply]
Oh sorry, I saw `--coverage`, `--coverage-output`, etc. in the docs regarding [command-line-flags](https://thomhurst.github.io/TUnit/docs/tutorial-extras/command-line-flags), and this led me to the wrong conclusion.

Indeed If I try from the CLI to run `dotnet run -c Release --report-trx --coverage` I see this error:
![image](https://github.com/user-attachments/assets/b3b2b7be-f364-4ab5-9c6b-0e105963928b)

The strange thing is that the part --report-trx --coverage is wrote also [here](https://thomhurst.github.io/TUnit/docs/tutorial-basics/running-your-tests#dotnet-run). 
I'm a bit confused right now.

**Reply by thomhurst** [reply]
Sorry, if you want to use coverage or trx you need to install the extension packages: https://thomhurst.github.io/TUnit/docs/extensions/
